src/platform_impl/linux/x11/event_processor.rs

```diff
@@ -1238,6 +1193,103 @@ impl<T: 'static> EventProcessor<T> {
         }
     }
 
+    fn handle_key_event<F>(
+        xev: &mut ffi::XEvent,
+        filtered_by_im: bool,
+        wt: &super::EventLoopWindowTarget<T>,
+        mod_keymap: &ModifierKeymap,
+        key_event_times: &mut [ffi::Time; 256],
+        device_mod_state: &mut ModifierKeyState,
+        active_window: Option<ffi::Window>,
+        callback: &mut F,
+    ) where
+        F: FnMut(Event<'_, T>),
+    {
+        use crate::event::ElementState::{Pressed, Released};
+
+        // Note that in compose/pre-edit sequences, this will always be Released.
+        let state = if xev.get_type() == ffi::KeyPress {
+            Pressed
+        } else {
+            Released
+        };
+
+        let xkev: &mut ffi::XKeyEvent = xev.as_mut();
+
+        let window = xkev.window;
+        let window_id = mkwid(window);
+
+        // Standard virtual core keyboard ID. XInput2 needs to be used to get a reliable
+        // value, though this should only be an issue under multiseat configurations.
+        let device = util::VIRTUAL_CORE_KEYBOARD;
+        let device_id = mkdid(device);
+        let keycode = xkev.keycode;
+
+        // When a compose sequence or IME pre-edit is finished, it ends in a KeyPress with
+        // a keycode of 0.
+        if keycode != 0 {
+            let signed_time = xkev.time as i64;
+            let time_since_previous_event =
+                signed_time.wrapping_sub(key_event_times[keycode as usize] as i64);
+
+            // An event is a replay of a previous event if it:
+            // 1) Isn't filtered by the IM.
+            // 2) Isn't the first time that an event with this keycode arrives.
+            // 3) Has a timestamp earlier or equal to the latest non-replayed event with this keycode.
```

Comment:
could maybe also be good for future readers to have some comments about why it's necessary to filter out these repeat events. This is kind of a hacky workaround really, so it feels like that should be really clearly sign posted in the code.